src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,222 @@
+  https://bugs.openjdk.java.net/browse/JDK-4509080."
+  (:require
+   [clojure.java.io :as io]
+   [nrepl.misc :refer [log]])
+  (:import
+   (clojure.lang Reflector)
+   (java.io BufferedInputStream BufferedOutputStream File)
+   (java.net InetSocketAddress ServerSocket Socket StandardProtocolFamily URI)
+   (java.nio ByteBuffer)
+   (java.nio.file Path)
+   (java.nio.channels Channels ClosedChannelException
+                      ServerSocketChannel SocketChannel)))
+
+(def orig-warn-on-reflection *warn-on-reflection*)
+
+(defmacro find-class [full-path]
+  `(try
+     (import ~full-path)
+     (catch ClassNotFoundException ex#
+       nil)))
+
+;;; InetSockets (TCP)
+
+(defn inet-socket [bind port]
+  (let [port (or port 0)
+        addr (fn [^String bind ^Integer port] (InetSocketAddress. bind port))
+        ;; We fallback to 127.0.0.1 instead of to localhost to avoid
+        ;; a dependency on the order of ipv4 and ipv6 records for
+        ;; localhost in /etc/hosts
+        bind (or bind "127.0.0.1")]
+    (doto (ServerSocket.)
+      (.setReuseAddress true)
+      (.bind (addr bind port)))))
+
+;; Unix domain sockets
+
+(def ^Class junixsocket-address-class
+  (find-class 'org.newsclub.net.unix.AFUNIXSocketAddress))
+
+(def ^Class junixsocket-server-class
+  (find-class 'org.newsclub.net.unix.AFUNIXServerSocket))
+
+(def ^Class jdk-unix-address-class
+  (find-class 'java.net.UnixDomainSocketAddress))
+
+(def ^Class jdk-unix-server-class
+  (find-class 'java.nio.channels.ServerSocketChannel))
+
+(def ^:private test-junixsocket?
+  ;; Make it possible to test junixsocket even when JDK >= 16
+  (= "true" (System/getProperty "nrepl.test.junixsocket")))
+
+(def ^:const unix-domain-flavor
+  (cond
+    test-junixsocket? (do
+                        (assert junixsocket-address-class)
+                        (assert junixsocket-server-class)
+                        (binding [*out* *err*]
+                          (println "nrepl.test: insisting on junixsocket support"))
+                        :junixsocket)
+    (and jdk-unix-address-class jdk-unix-server-class) :jdk
+    (and junixsocket-address-class junixsocket-server-class) :junixsocket
+    :else nil))
+
+(defn unix-socket-address
+  "Returns a filesystem socket address for the given path string."
+  [^String path]
+  (case unix-domain-flavor
+    :jdk
+    (Reflector/invokeStaticMethod jdk-unix-address-class "of"
+                                  ^"[Ljava.lang.String;" (into-array String [path]))
+    :junixsocket
+    (Reflector/invokeConstructor junixsocket-address-class
+                                 (into-array File [(File. path)]))
+    (let [msg "Support for filesystem sockets requires JDK 16+ or a junixsocket dependency"]
+      (log msg)
+      (throw (ex-info msg {:nrepl/kind ::no-filesystem-sockets})))))
+
+(set! *warn-on-reflection* false)
+
+(defn- bind [sock addr] (.bind sock addr))
+(defn- get-path [addr] (.getPath addr))
```

Comment:
Could you summarize why reflection is used here? What are the classes that can go down this path? Do they implement a common interface or not?

Reply:
Looks like we can drop `bind`, but I think we'll want to keep `get-path` (or something equivalent) since that method doesn't exist for `ServerSocket` before JDK 16.

Reply:
(...and of course AFUnixDomainSocket doesn't exist unless someone's added junixsocket it to their dependencies, and it's getPath returns String rather than Path.)